src/tiled/objectselectiontool.cpp

```diff
@@ -1431,20 +1433,45 @@ void ObjectSelectionTool::saveSelectionState()
 void ObjectSelectionTool::refreshCursor()
 {
     Qt::CursorShape cursorShape = Qt::ArrowCursor;
-
+    const MapObject *prevObject = nullptr;
     switch (mAction) {
     case NoAction: {
         const bool hasSelection = !mapScene()->selectedObjectItems().isEmpty();
 
         if ((mHoveredObjectItem || ((mModifiers & Qt::AltModifier) && hasSelection && !mHoveredHandle)) &&
                 !(mModifiers & Qt::ShiftModifier)) {
             cursorShape = Qt::SizeAllCursor;
+            MapObject *currentObject = mHoveredObjectItem->mapObject();
+
+            if (currentObject != prevObject) {
+                prevObject = currentObject;
+                if (highlightState)
+                    mapScene()->removeItem(mSelectionRectangle);
+
+                MapRenderer *renderer = mapDocument()->renderer();
+                QRectF boundingRect = objectBounds(currentObject, renderer,
+                                                   objectTransform(currentObject, renderer));
+
+                mSelectionRectangle->setRectangle(boundingRect);
```

Comment:
I think it's clever to reuse the selection rectangle, but it doesn't really do what I'd like to see. There is another class, the `MapObjectOutline` in `objectselectionitem.cpp` that I think would be more suitable because it matches the position as well as rotation of the object. You could define a new class `MapObjectHighlight` in `objectselectiontool.cpp` as a copy of `MapObjectOutline` and tweak it's `paint` override to look like the one of the `SelectionRectangle` class.

Later on, I think I'd like to change the look of the hover highlight, but that would suffice for now.